internal/cortex/frontend/transport/utils/utils.go

```diff
@@ -0,0 +1,158 @@
+		lruCache = nil
+		err = fmt.Errorf("Failed to create lru cache: %s", err)
+		return nil, err
+	}
+	return &FailedQueryCache{
+		regex:        regex,
+		errorExtract: errorExtract,
+		lruCache:     lruCache}, err
+}
+
+// UpdateFailedQueryCache returns true if query is cached so that callsite can increase counter, returns message as a string for callsite to log outcome
+func (f *FailedQueryCache) updateFailedQueryCache(err error, queryExpressionNormalized string, queryExpressionRangeLength int, lruCache *lru.Cache) (bool, string) {
+	// Extracting error code from error string.
+	codeExtract := f.errorExtract.FindStringSubmatch(err.Error())
+
+	// Checking if error code extracted successfully.
+	if codeExtract == nil || len(codeExtract) < 2 {
+		message := fmt.Sprintf(
+			`%s: %s, %s: %s, %s: %d, %s: %s`,
+			"msg", "String regex conversion error",
+			"normalized query", queryExpressionNormalized,
+			"query range seconds", queryExpressionRangeLength,
+			"updating cache for error", err,
+		)
+		return false, message
+	}
+
+	// Converting error code to int.
+	errCode, strConvError := strconv.Atoi(codeExtract[1])
+
+	// Checking if error code extracted properly from string.
+	if strConvError != nil {
+		message := fmt.Sprintf(
+			`%s: %s, %s: %s, %s: %d, %s: %s`,
+			"msg", "String to int conversion error",
+			"normalized query", queryExpressionNormalized,
+			"query range seconds", queryExpressionRangeLength,
+			"updating cache for error", err,
+		)
+		return false, message
+	}
+
+	// If error should be cached, store it in cache.
+	if !isCacheableError(errCode) {
+		message := fmt.Sprintf(
+			`%s: %s, %s: %s, %s: %d, %s: %s`,
+			"msg", "Query not cached due to non-cacheable error code",
+			"normalized query", queryExpressionNormalized,
+			"query range seconds", queryExpressionRangeLength,
+			"updating cache for error", err,
+		)
+		return false, message
+	}
+
+	// Checks if queryExpression is already in cache, and updates time range length value to min of stored and new value.
+	if contains, _ := lruCache.ContainsOrAdd(queryExpressionNormalized, queryExpressionRangeLength); contains {
+		if oldValue, ok := lruCache.Get(queryExpressionNormalized); ok {
+			queryExpressionRangeLength = min(queryExpressionRangeLength, oldValue.(int))
+		}
+		lruCache.Add(queryExpressionNormalized, queryExpressionRangeLength)
+	}
+
+	message := fmt.Sprintf(
+		`%s: %s, %s: %s, %s: %d, %s: %s`,
+		"msg", "Cached a failed query",
+		"normalized query", queryExpressionNormalized,
+		"range seconds", queryExpressionRangeLength,
+		"updating cache for error", err,
+	)
+	return true, message
+}
+
+// QueryHitCache checks if the lru cache is hit and returns whether to increment counter for cache hits along with appropriate message.
+func queryHitCache(queryExpressionNormalized string, queryExpressionRangeLength int, lruCache *lru.Cache) (bool, string) {
+	if value, ok := lruCache.Get(queryExpressionNormalized); ok && value.(int) <= queryExpressionRangeLength {
```

Comment:
The condition should be value.(int) <= queryExpressionRangeLength since queryExpressionRangeLength is the range length of the new query and value.(int) is the range length of the old query. From my understanding, it should only hit the cache and block the query if the old query's range length <= the new query's range length (new query is longer range length). 

I have a test case showing that it does not hit the cache here: [Test File Line 153](https://github.com/databricks/thanos/blob/0f29f88065d4abba0db790aaa05b7a1039a6e21c/internal/cortex/frontend/transport/utils/utils_test.go#L153)

Reply:
You are right.